### PR TITLE
UI Automation in Windows Console: only enable UIA when available

### PR DESCRIPTION
**Pull request by codeofdusk**
### Link to issue number:
Split from #9646 (builds on #9614).

### Summary of the issue:
Currently, in consoles with UI Automation enabled:

* If the advanced setting "force UI Automation in the Windows Console" is enabled on a system without a UIA console (such as Windows versions below Windows 10 1709 or those with the "use legacy console" option selected), the console is completely inaccessible.

### Description of how this pull request fixes the issue:
The legacy console code has been moved to `NVDAObjects.IAccessible.winConsoleLegacy`, and the check for consoles has been moved from `NVDAObjects.window.findOverlayClasses` to `NVDAObjects.IAccessible.findOverlayClasses`. Since NVDA prefers UIA over MSAA, UIA is used when enabled and available and legacy is used otherwise.

### Testing performed:
Tested all combinations of the "use legacy console" and UIA checkboxes on Windows 10 versions 1803 and 1903.

### Known issues with pull request:
See #9646 for all UIA console known issues to date.

### Change log entry:
None.

### CODE REVIEW

**Comment by codeofdusk** [comment]
CC @feerrenrut, @leonardder, @michaelDCurran.

**Comment by leonardder** [comment]
> The legacy console code has been moved to `NVDAObjects.IAccessible.winConsoleLegacy`.

Why do you consider this necessary? This breaks all add-ons that somehow use NVDAObjects.window.winConsole.

> and the check for consoles has been moved from `NVDAObjects.window.findOverlayClasses` to `NVDAObjects.IAccessible.findOverlayClasses`.

This change makes much sense.

File: source/core.py [hunk 1]
```diff
@@ -416,9 +416,9 @@ def handlePowerStatusChange(self):
 		log.warning("Java Access Bridge not available")
 	except:
 		log.error("Error initializing Java Access Bridge support", exc_info=True)
-	import winConsoleHandler
-	log.debug("Initializing winConsole support")
-	winConsoleHandler.initialize()
+	import winConsoleHandlerLegacy
```
Comment (leonardder) [inline source/core.py]: Why did you rename this as well?

**Comment by michaelDCurran** [comment]
Legacy winConsoles must be handled by IAccessible now so that order of 
choice can go UIA, then IAccessible. I think it is logical that 
WinConsoleLegacy be stored in NVDAObjects.IAccessible, but I do 
understand that this could break add-ons.

Would it be good enough to add back an NVDAObjects.window.winconsole 
module that imports NVDAObjects.IAccessible.winConsoleLegacy as 
WinConsole, and also logs a deprecation warning?

File: source/core.py (hunk 1 above)
Comment (codeofdusk) [reply source/core.py]: To avoid confusion (in anticipation of UIA becoming the default).

File: source/NVDAObjects/IAccessible/__init__.py [hunk 2]
```diff
@@ -528,6 +528,9 @@ def findOverlayClasses(self,clsList):
 		elif windowClassName.startswith("Chrome_"):
 			from . import chromium
 			chromium.findExtraOverlayClasses(self, clsList)
+		elif windowClassName == "ConsoleWindowClass":
```
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/__init__.py]: Rather than placing the code directly here, you can just place an entry in the _staticMap directory in this file. But as well as the window class, also  check for an MSAA role of ROLE_SYSTEM_CLIENT. This will ensure that the subclass is only used for the actual client of the window, and not the system menu or scrollbars etc.
 Something like:
```
("ConsoleWindowClass",oleacc.ROLE_SYSTEM_CLIENT):"winConsoleLegacy.WinConsoleLegacy",
```

File: source/NVDAObjects/IAccessible/winConsoleLegacy.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+#See the file COPYING for more details.
+#Copyright (C) 2007-2019 NV Access Limited, Bill Dengler
+
+import winConsoleHandlerLegacy as winConsoleHandler
+from . import IAccessible
+from ..behaviors import Terminal, EditableTextWithoutAutoSelectDetection
+import api
+import core
+
+class winConsoleLegacy(Terminal, EditableTextWithoutAutoSelectDetection, IAccessible):
```
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/winConsoleLegacy.py]: As this is a class, please use a capital W, I.e. WinConsoleLegacy.

File: source/NVDAObjects/window/winConsole.py [hunk 4]
```diff
@@ -1,71 +1,11 @@
...
-		"kb:alt+f4":"close",
-	}
+# NVDAObjects/window/winConsole.py
+# A part of NonVisual Desktop Access (NVDA)
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+# Copyright (C) 2019 Bill Dengler
+
+import warnings
+from ..IAccessible.winConsoleLegacy import winConsoleLegacy as WinConsole
```
Comment (michaelDCurran) [inline source/NVDAObjects/window/winConsole.py]: When capitalizing WinConsoleLegacy, don't forget to change it here.

File: source/NVDAObjects/window/winConsole.py [hunk 5]
```diff
@@ -1,71 +1,11 @@
...
+
+warnings.warn("NVDAObjects.window.winConsole is deprecated. Use NVDAObjects.IAccessible.winConsoleLegacy or NVDAObjects.UIA.winConsoleUIA instead",
+	DeprecationWarning, stacklevel=3)
```
Comment (michaelDCurran) [inline source/NVDAObjects/window/winConsole.py]: Please add a blank line at the end of this file.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
Again like with NVDAObjects.window.winConsole, can you please provide modules for winConsoleHandler and winCon with warnings when imported?
At the top of each you can just import everything from their respective new module. E.g for wincon:
```
from winCon import *
```

**Comment by codeofdusk** [comment]
@michaelDCurran Interestingly, this one merged without conflicts!

**Comment by michaelDCurran** [comment]
like @leonardder  I am a little concirned having so many modules renamed, even though of course we do have backwards compatibility wrapper modules with deprecation warnings. 
We are trying to keep changes on master fairly small, at least as far as breaking add-ons is concirned.
I think that it might be better if we don't rename winCon and winConsoleHandler. And rather than moving all of NVDAObjects.window.winConsole to NVDAObjects.IAccessible.winConsoleLegacy, we just create an NVDAObjects.IAccessible.winConsole which contains one class which is WinConsole that inherits from both NVDAObjects.window.winConsole.WinConsole and IAccessible. Very much like what I did with winWord.
Failing this, we could merge this pr straight to threshold, making the required renames, but in that case we wouldn't have to worry about the backwards compatibility modules.
But then master would not be able to take advantage of UIAWinConsole working instantly after turning it on in the Advanced settings pannel.
I'd like to get an opinion from @leonardder  on this. 
Sorry for not being as clear on this previously.

**Comment by leonardder** [comment]
I strongly second Micks proposal. Lets aim for master and revert the renames. Even on Threshold, these names are not required. I think it is enough to state in the module doc strings that this with a legacy implementation. Having said that, what's the definition of legacy? Even on Windows Server 2016, the legacy code will still be the used one, so I think this code will stick around for many years to come.

**Comment by codeofdusk** [comment]
@leonardder @michaelDCurran Is this better?

**Review by michaelDCurran — APPROVED** [review]
